### ISSUE TITLE
Sanity Check - Unicode Mismatch

### ISSUE THREAD

**Issue by piotroxp** [jrmuizel/pdf-extract#61]
I have created a PDF search application that scours your folders in search of documents and allows you to find keywords in the document.

At first,  I was not using this crate, but at some point it turned out that my app was not finding the right wording in the PDFs.  https://github.com/piotroxp/pdfscan

I am learning Rust at the same time when solving my real life need, which is going over terabytes of scientific PDF articles and finding the keywords in them.

Since I want to build a warp drive xD and have a very admirable cache of papers, you can understand that its critical for me to read all files regardless of encoding.

Today marks about 4 hours spent on looking at this error:

````
Unicode mismatch
````

For some PDF docs, it works. For others, mainly those downloaded from popular scientific publishers, i am hit with that log.

My repo is attached just so you can understand what I want to achieve.

Wherein is the issue? I am new to Rust. I'm pretty sure that Rust, being a systems programming language, does supply PDF libs regardless of encoding. I can be wrong in that statement.

How can I fix my code? Ideally, I would enjoy the ability to read in bytes raw, and only then transform that representation to utf8. Right now, I am unable to search through sci papers.

This ticket is created just because I find it amusing and mentally challenging to understand what I do wrong. Unless you are doing something wrong, which is also a learning expierience.

**Comment by jrmuizel** [comment]
Can you share a PDF that this happens with?

**Comment by piotroxp** [comment]
Sending over some example errors I'm running into along with the files @jrmuizel 

[IMX8QMAEC-1815333.pdf](https://github.com/jrmuizel/pdf-extract/files/11853107/IMX8QMAEC-1815333.pdf)
````
Thread started to search in: /nvme2/test/unicode
/nvme2/test/unicode/IMX8QMAEC-1815333.pdf
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
thread '<unnamed>' panicked at 'missing char 49 in map {32: "X", 12: "V", 1: "1", 29: "P", 21: "C", 26: "L", 36: "Y", 34: "G", 3: "3", 7: "7", 15: "M", 23: "E", 14: "_", 17: "N", 2: "2", 5: "5", 24: "T", 6: "6", 19: "D", 8: "8", 11: "A", 18: "U", 37: "Z", 25: "B", 33: "O", 4: "4", 28: "K", 10: "0", 35: "J", 16: "I", 9: "9", 20: "H", 13: "S", 31: "Q", 30: "W", 27: "F", 22: "R"} for <</BaseFont /QCGLDF+ArialMT/Encoding /WinAnsiEncoding/FirstChar 32/FontDescriptor 831 0 R/LastChar 95/Subtype /TrueType/ToUnicode 833 0 R/Type /Font/Widths [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 556 556 556 556 556 556 556 556 556 556 0 0 0 0 0 0 0 666 666 722 722 666 610 777 722 277 500 666 556 833 722 777 666 777 722 666 610 722 666 943 666 666 610 0 0 0 0 556]>>', /home/piotro/.cargo/registry/src/github.com-1ecc6299db9ec823/pdf-extract-0.6.5/src/lib.rs:746:27
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Search results have been zipped to: search_results_20230623211811.zip
````


[Bloom.pdf](https://github.com/jrmuizel/pdf-extract/files/11853114/Bloom.pdf)
````
Thread started to search in: /nvme2/test/threaderror/
/nvme2/test/threaderror/Bloom.pdf
thread '<unnamed>' panicked at 'no widths', /home/piotro/.cargo/registry/src/github.com-1ecc6299db9ec823/pdf-extract-0.6.5/src/lib.rs:570:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Search results have been zipped to: search_results_20230623211904.zip
````

[qt7vq3z6v1_noSplash_11cf93c4e513781acd1abae3cbe4e90d.pdf](https://github.com/jrmuizel/pdf-extract/files/11853121/qt7vq3z6v1_noSplash_11cf93c4e513781acd1abae3cbe4e90d.pdf)

````
/nvme2/test/other1/qt7vq3z6v1_noSplash_11cf93c4e513781acd1abae3cbe4e90d.pdf
thread '<unnamed>' panicked at 'missing char 104 in map {217: "Ψ", 56: "\u{f8f1}", 50: "\u{f8ee}", 55: "\u{f8fa}", 208: "Γ", 218: "Ω", 66: "\u{f8ec}", 159: "√", 211: "Λ", 64: "\u{f8ed}", 213: "Π", 60: "\u{f8f2}", 210: "Θ", 62: "\u{f8f4}", 214: "Σ", 57: "\u{f8fc}", 160: " ", 212: "Ξ", 216: "Φ", 54: "\u{f8ef}", 58: "\u{f8f3}", 209: "∆", 61: "\u{f8fd}", 48: "\u{f8eb}", 215: "Υ", 59: "\u{f8fe}", 51: "\u{f8f9}", 52: "\u{f8f0}", 67: "\u{f8f7}", 63: "\u{f8e6}", 65: "\u{f8f8}", 49: "\u{f8f6}", 53: "\u{f8fb}"} for <</BaseFont /BJDRNW+CMEX10/FirstChar 0/FontDescriptor 1542 0 R/LastChar 125/Subtype /Type1/ToUnicode 332 0 R/Type /Font/Widths 1523 0 R>>', /home/piotro/.cargo/registry/src/github.com-1ecc6299db9ec823/pdf-extract-0.6.5/src/lib.rs:746:27
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Search results have been zipped to: search_results_20230623212029.zip
````

**Comment by sagarp-patel** [comment]
Hi everyone, I am getting a unicode mismatch error too. It results the entire program quitting via panic. Is it possible if we can figure out if a pdf is suitable to work with the crate or not before hand, or instead of panicking is it possible to return an error, so that we can keep processing the rest of the files?? 
Here is what I got for my error 
Unicode mismatch
Unicode mismatch
Unicode mismatch
Unicode mismatch
thread '<unnamed>' panicked at 'unexpected smask type 251 0 R',\pdf-extract-0.6.5\src\lib.rs:1209:24
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

**Comment by piotroxp** [comment]
@sagarp-patel probably the pdf file would be good as well

**Comment by sagarp-patel** [comment]
@piotroxp I think I found a solution that works for now. You can handle the panics using std::panic::catch_unwind . This will also keep the data from what has already been processed in the pdf so far. So just do a 
```match std::panic::catch_unwind(move || {
    \\Your PDF processing code that panics goes here
}){
Ok(data)=>{handle whatever data was processed},
Err(err)=>{handle error}
}
```

**Comment by saravmajestic** [comment]
i have similar error using this PDF: https://www3.weforum.org/docs/WEF_Future_of_Jobs_2020.pdf
```
Unicode mismatch true fi "fi" Ok("ﬁ") [64257]
Unicode mismatch true fl "fl" Ok("ﬂ") [64258]
Unicode mismatch true fi "fi" Ok("ﬁ") [64257]
Unicode mismatch true fi "fi" Ok("ﬁ") [64257]
thread 'main' panicked at 'color_space [67, 83, 48] "DeviceN" [/DeviceN, [/Black], /DeviceCMYK, 8390 0 R, 8393 0 R]', /workspace/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pdf-extract-0.7.2/src/lib.rs:1428:25
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by 0xricksanchez** [comment]
[Another PDF](https://wcventure.github.io/pdf/ICSE2020_MemLock.pdf) that reproduces this. Would be amazing if we can get this library on a level of [markitdown](https://github.com/microsoft/markitdown) :heart:

**Comment by jrmuizel** [comment]
WEF_Future_of_Jobs_2020.pdf is fixed by 47898583ce0ab32b3fc4b1006395f716a0557aae

**Comment by jrmuizel** [comment]
I checked the other PDF's in this thread and it looks like none of them panic anymore. If I missed any or you find new ones that do panic please open a new issue.
